datastore.go

```diff
@@ -130,6 +131,47 @@ func DiskUsage(d Datastore) (uint64, error) {
 	return persDs.DiskUsage()
 }
 
+// TTLDatastore is an interface that should be implemented by datastores that
+// support expiring entries.
+type TTLDatastore interface {
+	Datastore
+
+	PutWithTTL(key Key, value interface{}, ttl time.Duration) error
+	SetTTL(key Key, ttl time.Duration) error
+}
+
+// Txn is an interface for transactions that can be committed or discarded.
+type Txn interface {
+	// Txn has the same capabilities as Datastore. Txns allow users to batch
+	// queries and mutations to the Datastore into atomic groups, or transactions.
+	// Actions performed on a transaction will not take hold until a successful
+	// call to Commit has been made. Likewise, transactions can be aborted by
+	// calling Discard before a successful Commit has been made.
+	Put(key Key, value interface{}) error
+	Get(key Key) (value interface{}, err error)
+	Has(key Key) (exists bool, err error)
+	Delete(key Key) error
+	Query(q query.Query) (query.Results, error)
```

Comment:
I'd rather extend `Datastore` here so we avoid duplicating these

Reply:
so i generally agree, though this change was made in response to discussion happening in the badger impl, specifically [right here](https://github.com/ipfs/go-ds-badger/pull/27#pullrequestreview-143979968). what do you think? the idea was to make it clear that this is happening in the `Txn` not directly on the `Datastore`. open to eitehr

Reply:
The rationale behind this is that `Txn` is not extending the `Datastore` interface but is actually redefining it (committing is not an optional new feature, it's mandatory for the `Put` to take effect). I agree that this solution is also not ideal, seems to me the lesser of the two evils, but I'm okay if you want to use `Datastore` here.

Reply:
further thoughts on this? think we're pretty close!